pkg/report/sarif.go

```diff
@@ -0,0 +1,190 @@
+	sw.run.Tool.Driver.WithVersion(sw.Version)
+	sw.run.Tool.Driver.WithFullName("Trivy Vulnerability Scanner")
+
+	ruleIndexes := map[string]int{}
+	for _, res := range report.Results {
+		for _, vuln := range res.Vulnerabilities {
+			fullDescription := vuln.Description
+			if fullDescription == "" {
+				fullDescription = vuln.Title
+			}
+			path := vuln.PkgPath
+			if path == "" {
+				path = res.Target
+			}
+
+			sw.addSarifResult(&sarifData{
+				vulnerabilityId:  vuln.VulnerabilityID,
+				severity:         vuln.Severity,
+				url:              vuln.PrimaryURL,
+				resourceType:     res.Type,
+				artifactLocation: toPathUri(path),
+				resultIndex:      getRuleIndex(vuln.VulnerabilityID, ruleIndexes),
+				fullDescription:  html.EscapeString(fullDescription),
+				helpText: fmt.Sprintf("Vulnerability %v\nSeverity: %v\nPackage: %v\nFixed Version: %v\nLink: [%v](%v)\n%v",
+					vuln.VulnerabilityID, vuln.Severity, vuln.PkgName, vuln.FixedVersion, vuln.VulnerabilityID, vuln.PrimaryURL, vuln.Description),
+				helpMarkdown: fmt.Sprintf("**Vulnerability %v**\n| Severity | Package | Fixed Version | Link |\n| --- | --- | --- | --- |\n|%v|%v|%v|[%v](%v)|\n\n%v",
+					vuln.VulnerabilityID, vuln.Severity, vuln.PkgName, vuln.FixedVersion, vuln.VulnerabilityID, vuln.PrimaryURL, vuln.Description),
+				message: fmt.Sprintf("Package: %v\nInstalled Version: %v\nVulnerability %v\nSeverity: %v\nFixed Version: %v\nLink: [%v](%v)",
+					vuln.PkgName, vuln.InstalledVersion, vuln.VulnerabilityID, vuln.Severity, vuln.FixedVersion, vuln.VulnerabilityID, vuln.PrimaryURL),
+			})
+		}
+		for _, misconf := range res.Misconfigurations {
+			sw.addSarifResult(&sarifData{
+				vulnerabilityId:  misconf.ID,
+				severity:         misconf.Severity,
+				url:              misconf.PrimaryURL,
+				resourceType:     res.Type,
+				artifactLocation: toPathUri(res.Target),
+				resultIndex:      getRuleIndex(misconf.ID, ruleIndexes),
+				fullDescription:  html.EscapeString(misconf.Description),
+				helpText: fmt.Sprintf("Misconfiguration %v\nType: %s\nSeverity: %v\nCheck: %v\nMessage: %v\nLink: [%v](%v)\n%s",
+					misconf.ID, misconf.Type, misconf.Severity, misconf.Title, misconf.Message, misconf.ID, misconf.PrimaryURL, misconf.Description),
+				helpMarkdown: fmt.Sprintf("**Misconfiguration %v**\n| Type | Severity | Check | Message | Link |\n| --- | --- | --- | --- | --- |\n|%v|%v|%v|%s|[%v](%v)|\n\n%v",
+					misconf.ID, misconf.Type, misconf.Severity, misconf.Title, misconf.Message, misconf.ID, misconf.PrimaryURL, misconf.Description),
+				message: fmt.Sprintf("Artifact: %v\nType: %v\nVulnerability %v\nSeverity: %v\nMessage: %v\nLink: [%v](%v)",
+					res.Target, res.Type, misconf.ID, misconf.Severity, misconf.Message, misconf.ID, misconf.PrimaryURL),
+			})
+		}
+	}
+	sarifReport.AddRun(sw.run)
+	return sarifReport.PrettyWrite(sw.Output)
+}
+
+func toSarifRuleName(vulnerabilityType string) string {
```

Comment:
We may be able to use `report.Result.Class`.

Reply:
Also, what about misconfigurations?

Reply:
yes, added config files.